src/bigint.rs

```diff
@@ -467,146 +546,352 @@ impl Unsigned for BigUint {}
-                + (carry as DoubleBigDigit);
-            let (hi, lo) = big_digit::from_doublebigdigit(d);
-            carry = hi;
-            *a = lo;
+        while self.data.len() < new_len {
+            self.data.push(0);
         }
 
-        if carry != 0 { sum.push(carry); }
-        BigUint::new(sum)
+        add2(&mut self.data[..], &other.data[..]);
+        self.normalize()
     }
 }
 
 forward_all_binop_to_val_ref!(impl Sub for BigUint, sub);
 
+fn sub2(a: &mut [BigDigit], b: BigUintS) {
+    let mut b_iter = b.iter();
+    let mut borrow = 0;
+
+    for ai in a.iter_mut() {
+        if let Some(bi) = b_iter.next() {
+            *ai = sbb(*ai, *bi, &mut borrow);
+        } else if borrow != 0 {
+            *ai = sbb(*ai, 0, &mut borrow);
+        } else {
+            break;
+        }
+    }
+
+    debug_assert!(borrow == 0 && b_iter.all(|x| *x == 0),
+            "Cannot subtract b from a because b is larger than a.");
+}
+
+fn sub_sign(a: BigUintS, b: BigUintS) -> BigInt {
+    fn cmp(a: BigUintS, b: BigUintS) -> Ordering {
+        /*
+         * Since we're working with slices here, it's _not_ guaranteed that the highest element in
+         * the slice is nonzero:
+         */
+        let mut i = cmp::max(a.len(), b.len());
+
+        while {
+            let av = if i < a.len() { a[i] } else { 0 };
+            let bv = if i < b.len() { b[i] } else { 0 };
+
+            if av < bv { return Less; }
+            if av > bv { return Greater; }
+
+            i != 0
+        } { i = i - 1; }
+
+        return Equal;
+    }
+
+    match cmp(a, b) {
+        Less    => {
+            let mut ret = BigUint::from_slice(b);
+            sub2(&mut ret.data[..], a);
+            BigInt::from_biguint(Plus, ret.normalize())
+        }
+        Greater => {
+            let mut ret = BigUint::from_slice(a);
+            sub2(&mut ret.data[..], b);
+            BigInt::from_biguint(Plus, ret.normalize())
+        }
+        _       => Zero::zero(),
+    }
+}
+
 impl<'a> Sub<&'a BigUint> for BigUint {
     type Output = BigUint;
 
-    fn sub(self, other: &BigUint) -> BigUint {
-        let mut diff = self.data;
-        let other = &other.data;
-        assert!(diff.len() >= other.len(), "arithmetic operation overflowed");
+    fn sub(mut self, other: &BigUint) -> BigUint {
+        sub2(&mut self.data[..], &other.data[..]);
+        self.normalize()
+    }
+}
 
-        let mut borrow: DoubleBigDigit = 0;
-        for (a, &b) in diff.iter_mut().zip(other.iter()) {
-            let d = big_digit::BASE - borrow
-                + (*a as DoubleBigDigit)
-                - (b as DoubleBigDigit);
-            let (hi, lo) = big_digit::from_doublebigdigit(d);
-            /*
-            hi * (base) + lo == 1*(base) + ai - bi - borrow
-            => ai - bi - borrow < 0 <=> hi == 0
-             */
-            borrow = if hi == 0 { 1 } else { 0 };
-            *a = lo;
-        }
+forward_all_binop_to_ref_ref!(impl Mul for BigUint, mul);
```

Comment:
Just related to this, since you reverted back to simple ref-ref, we could revert BigInt's Mul back to simple forwarding too, rather than the `bigint_mul` dance I had added.